docs/greenboot.md

```diff
@@ -0,0 +1,74 @@
+# Integrating MicroShift with Greenboot
+
+## Motivation
+
+Serviceability of Edge Devices is often limited or non-existent, which makes it
+challenging to troubleshoot device problems following a failed software or
+operating system upgrade. 
+
+To mitigate these problems, MicroShift uses [greenboot](https://github.com/fedora-iot/greenboot),
+the Generic Health Check Framework for `systemd` on `rpm-ostree` based systems.
+If a failure is detected, the system is expected to boot into the last known
+working configuration using `rpm-ostree` rollback facilities.
+
+This functionality benefits the users by reducing the risk of being locked out
+of an Edge Device when upgrades take place. Users should not experience significant
+interruption of service in case of a failed upgrade.
+
+## Implementation Details
+
+### MicroShift Validation
+
+The MicroShift core RPM package includes the `microshift-greenboot-required.sh`
+health check script to validate that all the required MicroShift services are
+up and running. There is no explicit dependency on the `greenboot` RPM. The
+script is installed into the `/etc/greenboot/check/required.d` directory and
+it is not executed during the system boot in case the `greenboot` package is
+not present.
+
+In addition, if the `greenboot-default-health-check` RPM subpackage is installed,
+it already includes
+[health check scripts](https://github.com/fedora-iot/greenboot#health-checks-included-with-subpackage-greenboot-default-health-checks)
+verifying that DNS and `ostree` services can be accessed.
+
+Exiting the script with a non-zero status will have the boot declared as failed.
+Greenboot redirects all the script output to the system log, accessible via the
+`journalctl -u greenboot-healthcheck.service` command.
+
+|Validation                                        |Pass  |Fail  |
+|--------------------------------------------------|------|------|
+|Check microshift.service is enabled               |Next  |exit 0|
+|Wait for microshift.service to be active (!failed)|Next  |exit 1|
+|Wait for any pods to start                        |Next  |exit 1|
+|For each core namespace, wait for pods to be ready|Next  |exit 1|
+|For each core namespace, check pods not restarting|exit 0|exit 1|
```

Comment:
```suggestion
|For each core namespace, check pods not restarting|exit 0|exit 1|
```

This check can be dangerous, some pods will restart, for example, pods that depend on microshift, will miss microshift during a shutdown, etc... and restart..

May be we will need to disable this at some point, or leave it out for the time being, maybe even white list some pods?

Reply:
@mangelajo, please look at the implementation PR for this check. It is done by comparing samples over time, so there should be more non-restarting than restarting samples